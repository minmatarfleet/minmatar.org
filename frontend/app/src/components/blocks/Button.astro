---
import FlexInline from '@components/compositions/FlexInline.astro';
import type { ButtonColors, ButtonSizes } from '@dtypes/LayoutComponents';

interface Props {
    color?: ButtonColors;
    external?: boolean;
    href?: string;
    size?: ButtonSizes;
    [propName: string]: any;
}

const {
    color = 'fleet-red',
    external = false,
    href,
    size = 'lg',
    ...attributes
} = Astro.props;

delete attributes.class

const color_prop = `var(--${color})`;

const Element = href ? 'a' : 'button';
---

<Element
    href={ href }
    class:list={['[ button ]', Astro.props.class]}
    {...attributes}
    data-size={size}
    style={`--color_prop: ${color_prop}`}
>
    <FlexInline gap="var(--space-xs)">
        <slot name="icon" />
        <slot />
    </FlexInline>
</Element>

<style lang="scss">

    button,
    a[href].button {
        --background-color: var(--color_prop);

        display: block;
        text-decoration: none;
        font-family: var(--button-font);
        text-transform: uppercase;
        text-align: left;
        padding: var(--space-s) var(--space-l-xl);
        border: solid 1px var(--background-color);
        color: var(--button-color);
        background-color: var(--background-color);
        align-items: center;
        transition: var(--fast-transition);

        &[disabled] {
            filter: saturate(50%);
        }

        &:focus-visible {
            outline: none;
        }
        
        /* Desktop */
        @media (hover: hover) {
            &:focus-visible,
            &:hover:not([disabled]) {
                cursor: pointer;
                transform: scale(1.05);
            }
        }
        
        &:active:not([disabled]) {
            transform: scale(1);
        }

        /* Tactile */
        @media (hover: none) {
            transform: scale(1);

            &:active:not([disabled]) {
                transform: scale(0.95);
            }
        }

        &[data-size="sm"] {
            padding: var(--space-2xs) var(--space-xs-s);
        }

        &.w-full > div {
            justify-content: center;
        }
    }
</style>