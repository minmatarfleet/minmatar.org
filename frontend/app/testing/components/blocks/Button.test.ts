import { experimental_AstroContainer as AstroContainer } from "astro/container";
import { expect, test } from "vitest";
import Button from "@components/blocks/Button.astro";

test("Button defaults", async () => {
  const container = await AstroContainer.create();
  const result = await container.renderToString(Button, {});

  expect(result).toMatchSnapshot();
});
