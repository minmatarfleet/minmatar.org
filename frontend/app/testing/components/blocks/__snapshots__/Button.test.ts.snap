// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button defaults 1`] = `"<button data-size="lg" style="--button_background_color: var(--fleet-red); --button_color: var(--fleet-yellow)" disabled x-init="$el.removeAttribute(&#34;disabled&#34;)" data-astro-cid-r5e2wo56="true" class="[ button ]"> <div class="flex-inline" data-astro-cid-r5e2wo56="true" data-astro-cid-ud5fqf5k style="--flex_inline_gap: var(--space-xs);--flex_inline_justification: center;">     </div>   </button> "`;
